test/unit/DominantJuice.unit.t.sol

```diff
@@ -0,0 +1,962 @@
+        vm.prank(address(ethPaymentTerminal));
+        dominantJuice.payParams(payParamsData);
+    }
+
+    // PASSING
+    function testFuzz_payParams_revertsWhenPaymentIsBelowMinPledge(uint256 _value) public bonusDeposited {
+        // In reality, payParams() will be called by the JB Payment Terminal Store, but mocking a direct call is
+        // a good check for unit testing purposes.
+        vm.assume(_value < MIN_PLEDGE_AMOUNT);
+        payParamsData.amount.value = _value;
+
+        vm.expectRevert(abi.encodeWithSelector(DominantJuice.AmountIsBelowMinimumPledge.selector, MIN_PLEDGE_AMOUNT));
+        vm.prank(rando);
+        dominantJuice.payParams(payParamsData);
+    }
+
+    // PASSING - (unimplemented feature error)
+    function test_payParams_happyPathReturnsMemoryVariables() public bonusDeposited {
+        payParamsData.payer = pledger1;
+        payParamsData.weight = 1;
+        payParamsData.memo = "juice";
+        payParamsData.amount = JBTokenAmount(ethToken, MIN_PLEDGE_AMOUNT, 18, 1);
+
+        // Error: Unimplemented feature (/solidity/libsolidity/codegen/ArrayUtils.cpp:228):
+        // Copying of type struct JBPayDelegateAllocation3_1_1 memory[] memory to storage not yet supported.
+        // delegateAllocations = new JBPayDelegateAllocation3_1_1[](1);
+        // delegateAllocations[0] = JBPayDelegateAllocation3_1_1(dominantJuice, 0, "");
+
+        (uint256 _weight, string memory _memo,) = dominantJuice.payParams(payParamsData);
+        assertEq(payParamsData.weight, _weight);
+        assertEq(payParamsData.memo, _memo);
+        //assertEq(0, delegateAllocations[0].amount);
+
+        // Assert that there are no major contract changes when payParams is called directly:
+        DominantJuice.FundingStatus memory status = dominantJuice.getCycleFundingStatus();
+
+        assertEq(0, status.totalPledged);
+        assertEq(0, status.percentOfGoal);
+        assertFalse(status.isTargetMet);
+        assertFalse(status.hasCycleExpired);
+        assertEq(TOTAL_REFUND_BONUS, dominantJuice.getBalance());
+        (uint256 pledgerW, uint256 totalW) = dominantJuice.getPledgerAndTotalWeights(pledger1);
+        assertEq(0, pledgerW);
+        assertEq(0, totalW);
+    }
+
+    // payParams() function is non-payable, the parent function that calls it is nonReentrant, and there are only if
+    // statements and logic to satisfy Juicebox architecture, thus not many unit tests here. This function can be
+    // called directly, but nothing would happen. The only way to pledge correctly is by calling
+    // `JBPayoutRedemptionPaymentTerminal3_1_1.pay()`.
+
+    ///////////////////////
+    // didPay() Tests
+    ///////////////////////
+
+    // PASSING
+    function test_didPay_revertsIfPaymentSent(uint256 _amount) public {
+        _amount = bound(_amount, 1, 2000 ether);
+        didPayData.projectId = projectID;
+
+        vm.expectRevert("Pledges should be made through JB website.");
+        vm.prank(rando);
```

Comment:
Because the `msg.value == 0` check happens before the terminal check, not at same time like it was previously, so that each one could have it's own custom error. Is that the question or something else?

Reply:
that's true. i was just saying that the test should only test one thing, and calling with rando is _another_ failure....just a later one. So if you happened to move the rando check first (vs. msg.value == 0) in the original contract code, this test will fail, but not because the logic is wrong, but just because the message was wrong....because it failed for a different error. It creates brittle tests.

Reply:
I see what you're saying. Yes, I agree and have changed rando to payment terminal, and I'll go through and see if there are any other cases like that as I make (hopefully) final passes this week.